src/commissioner/Events.h

```diff
@@ -0,0 +1,274 @@
+/*
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <controller/CommissioneeDeviceProxy.h>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <protocols/secure_channel/CASESession.h>
+#include <protocols/secure_channel/PASESession.h>
+#include <protocols/secure_channel/RendezvousParameters.h>
+#include <setup_payload/ManualSetupPayloadParser.h>
+#include <setup_payload/QRCodeSetupPayloadParser.h>
+#include <setup_payload/SetupPayload.h>
+
+namespace chip {
+namespace Commissioner {
+namespace Events {
+
+constexpr size_t kMaxAttestationElementsLength = 900; // RESP_MAX
+constexpr size_t kMaxNocsrElementsLength       = 900; // RESP_MAX
+constexpr size_t kMaxNetworkIdLength           = 32;
+
+struct SharedBuffer
+{
+    const ByteSpan Get() const { return ByteSpan(mData.get(), mLen); }
+    MutableByteSpan GetMutable() { return MutableByteSpan(mData.get(), mCapacity); };
+    CHIP_ERROR Set(const ByteSpan buffer)
+    {
+        if (mCapacity < buffer.size())
+        {
+            ReturnErrorOnFailure(Allocate(buffer.size()));
+        }
+        memmove(mData.get(), buffer.data(), buffer.size());
+        mLen = buffer.size();
+        return CHIP_NO_ERROR;
+    }
+    void Clear()
+    {
+        mData     = std::shared_ptr<uint8_t>(nullptr);
+        mCapacity = 0;
+        mLen      = 0;
+    }
+    CHIP_ERROR Allocate(size_t size)
+    {
+        mData = std::shared_ptr<uint8_t>(static_cast<uint8_t *>(chip::Platform::MemoryAlloc(size)), chip::Platform::MemoryFree);
```

Comment:
Use `Platform::SharedPtr` instead of `std::shared_ptr`?

Reply:
Ah, right.  I don't need to add anything to CHIPMem.h to do this.

OK, done.